### PR TITLE
Expose HTTP version of probe

### PR DESCRIPTION
**Pull request by Conorbro**
Implements #162.

Exposes HTTP version via `probe_http_version` gauge and probes can now have set valid HTTP versions that will fail if not met.

### CODE REVIEW

File: CONFIGURATION.md [hunk 1]
```diff
@@ -38,6 +38,9 @@ The other placeholders are specified separately.
   # Accepted status codes for this probe. Defaults to 2xx.
   [ valid_status_codes: <string>, ... | default = "2xx" ]
 
+  # Accepted HTTP versions for this probe.
+  [ valid_http_versions: <float64>, ... ]
```
Comment (brian-brazil) [inline CONFIGURATION.md]: I think a string might make more sense here

File: http.go [hunk 2]
```diff
@@ -85,12 +95,18 @@ func probeHTTP(target string, module Module, registry *prometheus.Registry) (suc
 			Name: "probe_ssl_earliest_cert_expiry",
 			Help: "Returns earliest SSL cert expiry in unixtime",
 		})
+
+		probeHTTPVersionGauge = prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "probe_http_version",
+			Help: "Returns the version of HTTP the probe used",
```
Comment (brian-brazil) [inline http.go]: It's about the response, not the request

File: http.go [hunk 3]
```diff
@@ -57,6 +58,15 @@ func matchRegularExpressions(reader io.Reader, httpConfig HTTPProbe) bool {
 	return true
 }
 
+func isValidHTTPVersion(probeVersion float64, validHTTPVersions []float64) bool {
```
Comment (brian-brazil) [inline http.go]: I'm not sure this needs to be a separate function

File: http.go [hunk 4]
```diff
@@ -180,6 +195,18 @@ func probeHTTP(target string, module Module, registry *prometheus.Registry) (suc
...
+
+		httpVersion, err := strconv.ParseFloat(strings.TrimPrefix(resp.Proto, "HTTP/"), 64)
+		if err == nil {
+			probeHTTPVersionGauge.Set(httpVersion)
+		} else {
+			log.Errorf("Error parsing HTTP version from probe response: %v", err)
+		}
+
+		if len(httpConfig.ValidHTTPVersions) != 0 {
+			success = isValidHTTPVersion(httpVersion, httpConfig.ValidHTTPVersions)
```
Comment (brian-brazil) [inline http.go]: This could override a previous success = false, only set if it this fails.

**Comment by Conorbro** [comment]
@brian-brazil

File: config.go [hunk 5]
```diff
@@ -35,16 +35,17 @@ type Module struct {
...
-	PreferredIPProtocol    string                  `yaml:"preferred_ip_protocol,omitempty"`
-	NoFollowRedirects      bool                    `yaml:"no_follow_redirects,omitempty"`
-	FailIfSSL              bool                    `yaml:"fail_if_ssl,omitempty"`
-	FailIfNotSSL           bool                    `yaml:"fail_if_not_ssl,omitempty"`
-	Method                 string                  `yaml:"method,omitempty"`
-	Headers                map[string]string       `yaml:"headers,omitempty"`
-	FailIfMatchesRegexp    []string                `yaml:"fail_if_matches_regexp,omitempty"`
-	FailIfNotMatchesRegexp []string                `yaml:"fail_if_not_matches_regexp,omitempty"`
-	Body                   string                  `yaml:"body,omitempty"`
+	ValidStatusCodes       []int                   `yaml:"valid_status_codes"`
```
Comment (brian-brazil) [inline config.go]: These should still be omitempty

File: http.go [hunk 6]
```diff
@@ -180,6 +186,31 @@ func probeHTTP(target string, module Module, registry *prometheus.Registry) (suc
 		if success && (len(httpConfig.FailIfMatchesRegexp) > 0 || len(httpConfig.FailIfNotMatchesRegexp) > 0) {
 			success = matchRegularExpressions(resp.Body, httpConfig)
 		}
+
+		if err == nil {
```
Comment (brian-brazil) [inline http.go]: What err is this?

File: example.yml [hunk 7]
```diff
@@ -3,6 +3,7 @@ modules:
     prober: http
     timeout: 5s
     http:
+      valid_http_versions: []
```
Comment (brian-brazil) [inline example.yml]: I'd include an example that made it clean the valid values were 1.1 etc.

**Comment by Conorbro** [comment]
@brian-brazil

**Comment by brian-brazil** [comment]
Thanks!

**Comment by cceckman** [comment]
Thanks a bunch!

**Pull request merged**